src/WebJobs.Script.WebHost/Diagnostics/BufferedConsoleWriter.cs

```diff
@@ -0,0 +1,151 @@
+        {
+            _consoleBuffer = Channel.CreateBounded<string>(new BoundedChannelOptions(bufferSize) { SingleReader = true, SingleWriter = false });
+            _writeEvent = WriteToConsoleBuffer;
+            _consoleBufferTimeout = consoleBufferTimeout;
+            _writeResetEvent = new ManualResetEvent(true);
+            _maxLinesPerWrite = bufferSize / SingleWriteBufferDenominator;
+
+            if (autoStart)
+            {
+                StartProcessingBuffer();
+            }
+
+            _exceptionhandler = exceptionHandler;
+        }
+
+        public void WriteHandler(string evt)
+        {
+            _writeEvent(evt);
+        }
+
+        private void WriteToConsoleBuffer(string evt)
+        {
+            if (!_consoleBuffer.Writer.TryWrite(evt))
+            {
+                _writeResetEvent.Reset();
+                bool waitFailed = !_writeResetEvent.WaitOne(_consoleBufferTimeout);
+
+                // If the wait failed, write to the console. Otherwise, try the writing again - if that fails, write to the console.
+                if (waitFailed || !_consoleBuffer.Writer.TryWrite(evt))
+                {
+                    Console.WriteLine(evt);
+                }
+            }
+        }
+
+        // internal only for testing - should only be used from the constructor, not intended for concurrent callers.
+        internal void StartProcessingBuffer()
+        {
+            // intentional no-op if the task is already running
+            if (_consoleBufferReadLoop == null || _consoleBufferReadLoop.IsCompleted)
+            {
+                _consoleBufferReadLoop = ProcessConsoleBufferAsync();
+            }
+        }
+
+        private async Task ProcessConsoleBufferAsync()
+        {
+            try
+            {
+                var builder = new StringBuilder();
+
+                while (await _consoleBuffer.Reader.WaitToReadAsync())
+                {
+                    if (_consoleBuffer.Reader.TryRead(out string line1))
+                    {
+                        _writeResetEvent.Set();
+
+                        // Can we synchronously read multiple lines?
+                        // If yes, use the string builder to batch them together into a single write
+                        // If no, just write the single line without using the builder;
+                        if (_consoleBuffer.Reader.TryRead(out string line2))
+                        {
+                            builder.AppendLine(line1);
+                            builder.AppendLine(line2);
+                            int lines = 2;
+
+                            while (lines < _maxLinesPerWrite && _consoleBuffer.Reader.TryRead(out string nextLine))
+                            {
+                                builder.AppendLine(nextLine);
+                                lines++;
+                            }
+
+                            _writeResetEvent.Set();
```

Comment:
You chose to signal the event after writing potentially many lines, as opposed to signaling after each time you've successfully read a new line and made space in the channel. Probably doesn't make a huge difference, but wouldn't signaling after each read unblock the writer faster?

Reply:
Good question - when I was iterating on this code I did run a couple of quick tests with both variations, and I was not able to find a significant difference either way, but my testing wasn't particularly indepth.

Without having any useful data, I ended up going with the current approach. Here's my rough thinking, split into two cases - environments where we have multiple cores, and environments where we have a single core or fractional core.

Multiple cores - chances of the buffer getting full in a multicore environment are low, because it would require ~1000 function executions to occur without the threadpool dispatching the thread for the buffer consumer. So if the buffer is not getting full, why pay the cost of setting the MRE every iteration?

Single/fractional core - if we cannot run multiple threads concurrently, what is the benefit of waking up a writing thread while the consumer is still running?